src/cr/cube/matrix.py

```diff
@@ -28,187 +31,110 @@ def matrix(cls, cube, dimensions, slice_idx):
+            for _, idx, column in sorted(
+                itertools.chain(
+                    (column.ordering for column in self._inserted_columns),
+                    (column.ordering for column in self._base_columns),
+                )
+            )
         )
 
     @lazyproperty
-    def rows_dimension(self):
-        return self._base_matrix.rows_dimension
-
-    @lazyproperty
-    def table_base(self):
-        return self._base_matrix.table_base
-
-    @lazyproperty
-    def table_margin(self):
-        return self._base_matrix.table_margin
-
-    @lazyproperty
-    def _column_order(self):
-        """Indexer value identifying columns in order, suitable for slicing an ndarray.
-
-        This value is a 1D ndarray of int column indices, suitable for indexing the
-        columns array to produce an ordered version.
-        """
-        # ---Specifying int type prevents failure when there are zero columns. The
-        # ---default type for ndarray is float, which is not valid for indexing.
-        return np.array(self.columns_dimension.display_order, dtype=int)
-
-    @lazyproperty
-    def _row_order(self):
-        """Indexer value identifying rows in order, suitable for slicing an ndarray.
+    def _assembled_rows(self):
+        """Sequence of row vectors including inserted rows.
 
-        This value is a 1D ndarray of int row indices, suitable for indexing the rows
-        array to produce an ordered version.
+        Each row vector also reflects any new elements introduced by inserted columns.
         """
-        # ---Specifying int type prevents failure when there are zero rows---
-        return np.array(self.rows_dimension.display_order, dtype=int)
-
-
-class _MatrixWithHidden(object):
-    """Matrix with hidden vectors removed.
+        opposing_insertions = self._inserted_columns
 
-    A vector can be hidden explicitly by the user, or it can be automatically hidden
-    when it is empty and the prune option for the dimension is selected.
-    """
-
-    # ---Note that hiding a vector requires not just removing that vector, but also
-    # ---the element the removed vector contributes to each of the *opposing* vectors.
-    # ---For example, hiding a row is removing that row-vector from `.rows`, but also
-    # ---removing an element from each column-vector in `.columns`.
-
-    def __init__(self, base_matrix):
-        self._base_matrix = base_matrix
-
-    @lazyproperty
-    def columns(self):
         return tuple(
-            _VectorAfterHiding(column, self._base_matrix.rows)
-            for column in self._base_matrix.columns
-            if not column.hidden
+            _AssembledVector(row, opposing_insertions, 0 if idx < 0 else idx)
+            for _, idx, row in sorted(
+                itertools.chain(
+                    (row.ordering for row in self._inserted_rows),
+                    (row.ordering for row in self._base_rows),
+                )
+            )
```

Comment:
@slobodan-ilic @ernestoarbitrio It smells to me like a bug that an assembled row that it an insertion row relies on its idx value to be 0. Why would it be that the the "intersection" pval would always be the first value of `.pvals` on the intersecting column? I suppose if that `.pvals` was always a sequence of exactly one value that would make sense of it, but it sounds a little odd to me and at the very least is obscure without a comment.

Reply:
This is probably indicative of needing two extractions (either as classes or as mechanisms within one class). The assembled row that is _not_ an insertion row itself _needs_ this index. The assembled row that _is_ an insertion row, doesn't need it (but currently has it defaulting to 0).

Reply:
Well, the interesting thing is that at least one test fails on expectation doesn't match) if you replace `0 if idx < 0 else idx` with `idx`. `idx` always has a legitimate value, it's just end-based (like -2) and indexes into the insertions-collection if the vector is an insertion.

I took a quick look, and it appears this `idx` value is directly involved in the `pvals` calculation for an insertion, which just seems odd to me. I can't think of what calculation would _need_ a col_idx but work correctly with the same value for all insertions.

Reply:
@ernestoarbitrio can you weigh in on this one? ^^^